site/_posts/2017-08-16-quilt-arrow.md

```diff
@@ -0,0 +1,85 @@
+
+## Docker Analogy
+We’re big users of Docker. As we started working with data packages, we realized we wanted to create and deploy them like we do with Docker containers. Docker makes it easy to build containers and saved versioned snapshots. We wanted to do the same with data packages so we modeled the command-line interface on Docker.
+
+## DataFrames for Tabular Data
+We chose to focus special attention on tabular data. For Python data scientists, the `pandas.DataFrame` is the workhorse for columnar data. Whether data are coming from CSVs, databases, or from a web API, the first step for us (and seemingly most data scientists) is usually to load it into Pandas. Our goal was to make DataFrames part of packages so users can skip right to analysis without worrying about getting data into and out of files.
+
+## Cross Platform Data
+We wanted Quilt packages to work across the core languages of data science. A DataFrame in a package should be accessible as a DataFrame in Python, R and Spark.
+
+# The Challenge: Efficient Cross-Platform Data
+For unstructured data and images, we knew we could simply wrap and transport files in their native format. The trick was how we would handle tabular data in a way that felt natural to users in any language. In other words, how could we store and deliver DataFrames efficiently on all platforms.
+
+## Storing Tabular Data in <b>Parquet</b>
+We chose [Apache Parquet][4] to store tabular data because of its efficiency and widespread support, especially in the Hadoop and Spark communities. Parquet stores data in column chunks packed into row groups. Each column can be encoded in its own format and compressed. Columnar data formats like Parquet save storage and bandwidth because (1) they tend to be more easily and effectively compressed and, (2) they make it easy to read only a subset of columns.  In addition to being commonly used in Hadoop and Spark applications, Parquet is readable by several popular open-source SQL in-place query engines, including Hive, Impala and PrestoDB.
+
+When we started, Parquet support in Python was limited. In fact, searching for an open-source library to read and write Parquet files was how we discovered the Arrow project in the first place. We tested two Parquet libraries, fastparquet and pyarrow. The two libraries performed similarly for serializing and deserializing Pandas DataFrames to and from Parquet and in early versions we included hooks for both.
+
+## Bringing Data Packages into Memory with <b>Arrow</b>
+In addition to a storage format, to present tabular data in a consistent, ready-to-analyze state across platforms, we need to define a common <b>in-memory</b> representation of tabular data in packages. Arrow is exactly that. With Arrow, tabular data can be packaged as Arrow tables, ensuring that they have the exact same presentation in Python, R and Spark.
```

Comment:
> exact same presentation in Python, R and Spark.

I agree this is a goal, but there are some technical hurdles here that can't be hand-waved away (e.g. R has more limited in-memory data types than Python or Spark)